akka-stream/src/main/scala/akka/stream/impl/ActorRefSource.scala

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2018-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import akka.actor.{ ActorRef, PoisonPill }
+import akka.annotation.InternalApi
+import akka.stream.OverflowStrategies._
+import akka.stream._
+import akka.stream.stage._
+
+import scala.annotation.tailrec
+
+object ActorRefSource {
+  case object EagerComplete
+
+  private sealed trait ActorRefStage { def ref: ActorRef }
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final class ActorRefSource[T](
+    maxBuffer: Int,
+    overflowStrategy: OverflowStrategy,
+    completionMatcher: PartialFunction[Any, Unit],
+    failureMatcher: PartialFunction[Any, Throwable])
+    extends GraphStageWithMaterializedValue[SourceShape[T], ActorRef] {
+  import ActorRefSource._
+
+  val out: Outlet[T] = Outlet[T]("actorRefSource.out")
+
+  override val shape: SourceShape[T] = SourceShape.of(out)
+
+  def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ActorRef) =
+    throw new IllegalStateException("Not supported")
+
+  private[akka] override def createLogicAndEagerMaterializedValue(
+      inheritedAttributes: Attributes,
+      eagerMaterializer: Materializer): (GraphStageLogic, ActorRef) = {
+    val stage: GraphStageLogic with StageLogging with ActorRefStage = new GraphStageLogic(shape) with StageLogging
+    with ActorRefStage {
+      override protected def logSource: Class[_] = classOf[ActorRefSource[_]]
+
+      private val buffer: Buffer[T] =
+        if (maxBuffer != 0)
+          Buffer(maxBuffer, eagerMaterializer)
+        else {
+          null // todo: for backwards compatibility only
+        }
+      private var isCompleting: Boolean = false
+
+      override def preStart(): Unit = {
+        super.preStart()
+        if (maxBuffer == 0)
+          log.warning("for backwards compatibility: maxBuffer of 0 will not be supported in the future") // warning for backwards compatibility
+      }
+
+      override protected def stageActorName: String =
+        inheritedAttributes.get[Attributes.Name].map(_.n).getOrElse(super.stageActorName)
+
+      val ref: ActorRef = getEagerStageActor(eagerMaterializer, poisonPillFallback = true) {
+        case (_, EagerComplete) ⇒
+          completeStage()
+        case (_, PoisonPill) ⇒
+          log.warning("for backwards compatibility: PoisonPill will note be supported in the future")
+          isCompleting = true
+          pump()
+        case (_, m) if failureMatcher.isDefinedAt(m) ⇒
+          failStage(failureMatcher(m))
+        case (_, m) if completionMatcher.isDefinedAt(m) ⇒
+          isCompleting = true
+          pump()
+        case (_, m: T) ⇒ // todo: requires classTag (breaks api)
```

Comment:
I don't think this will be possible to solve without API change, but otoh the old ActorRefSourceActor did also not deal with the wrong types of elements, so maybe just accept that for now.

Reply:
Adding the class tag would break the API, but would be good to have at some point for more correctness because the API is in some sense lying now.

Reply:
It may solve itself when we get to Akka typed support for it, so I think that is fine.